### PR TITLE
Refactor settings controllers

### PR DESCRIPTION
**Pull request by Gargron**
- Disallow suspended accounts from revoking sessions and apps
- Allow suspended accounts to access exports

### CODE REVIEW

**Comment by Gargron** [comment]
I believe there might be an issue with allowing all non-functional (including "pending") accounts to access profile settings and import, so hold off on merging...

**Comment by SuperSandro2000** [comment]
> Disallow suspended accounts from revoking authorized applications, browser sessions, or changing 2FA settings

Why tough? If you are suspended you should still be allowed to revoke access from everything.

**Comment by Gargron** [comment]
If you are suspended, we retain some data to prevent ban evasions, e.g. you can’t sign up with the identical email address anymore. IPs and used apps may provide further insight into finding alternative accounts, so it doesn’t make sense that the suspended user can simply clean up all the tracks.

**Comment by SuperSandro2000** [comment]
> you can’t sign up with the identical email address anymore.

It's the part after the + filtered?

IPs are most of the time pretty useless. Mine for example changes every day and I can easily switch it.

> it doesn’t make sense that the suspended user can simply clean up all the tracks.

If my account got suspended by action taken on another device which maybe got stolen or the session cookie was phished etc. I can't revoke that device. I should be able to revoke those sessions without cleaning up all my tracks.

Also revoking or disabling 2FA does not clean up any track or provides any useful information to anyone.

**Comment by Gargron** [comment]
Why should you be able to change 2fa settings if you’re not allowed to change your password or email? It makes more sense to me to keep it consistent: suspended means the account’s done, no further changes.

**Comment by ThibG** [comment]
Before I actually review: I see no issue with being able to change 2FA settings or password when suspended. It would also be good to be able to revoke active sessions and change e-mail address, but I also understand how this is tied with the anti-abuse features so I think that's an acceptable compromise.

File: app/controllers/oauth/authorized_applications_controller.rb [hunk 1]
```diff
@@ -5,6 +5,7 @@ class Oauth::AuthorizedApplicationsController < Doorkeeper::AuthorizedApplicatio
 
   before_action :store_current_location
   before_action :authenticate_resource_owner!
+  before_action :require_not_suspended!
```
Comment (ThibG) [inline app/controllers/oauth/authorized_applications_controller.rb:8]: Could this be done only for the `destroy` action?

File: app/controllers/settings/featured_tags_controller.rb [hunk 2]
```diff
@@ -1,9 +1,6 @@
 # frozen_string_literal: true
 
 class Settings::FeaturedTagsController < Settings::BaseController
-  layout 'admin'
-
-  before_action :authenticate_user!
   before_action :set_featured_tags, only: :index
```
Comment (ThibG) [inline app/controllers/settings/featured_tags_controller.rb:4]: Hm, wouldn't it be consistent to require the account not to be suspended, here?

File: app/controllers/settings/identity_proofs_controller.rb [hunk 3]
```diff
@@ -1,9 +1,6 @@
 # frozen_string_literal: true
 
 class Settings::IdentityProofsController < Settings::BaseController
-  layout 'admin'
-
-  before_action :authenticate_user!
```
Comment (ThibG) [inline app/controllers/settings/identity_proofs_controller.rb:6]: Hm, wouldn't it be consistent to require the account not to be suspended, here?

File: app/controllers/settings/imports_controller.rb [hunk 4]
```diff
@@ -1,9 +1,6 @@
 # frozen_string_literal: true
 
 class Settings::ImportsController < Settings::BaseController
-  layout 'admin'
-
-  before_action :authenticate_user!
```
Comment (ThibG) [inline app/controllers/settings/imports_controller.rb:6]: Hm, wouldn't it be consistent to require the account not to be suspended, here?

File: app/controllers/settings/sessions_controller.rb [hunk 5]
```diff
@@ -1,11 +1,11 @@
 # frozen_string_literal: true
 
 class Settings::SessionsController < Settings::BaseController
-  before_action :authenticate_user!
-  before_action :set_session, only: :destroy
-
   skip_before_action :require_functional!
 
+  before_action :require_not_suspended!
```
Comment (ThibG) [inline app/controllers/settings/sessions_controller.rb:6]: could it be set only for `destroy`?

File: app/controllers/settings/two_factor_authentication/confirmations_controller.rb [hunk 6]
```diff
@@ -5,14 +5,12 @@ module TwoFactorAuthentication
     class ConfirmationsController < BaseController
       include ChallengableConcern
 
-      layout 'admin'
+      skip_before_action :require_functional!
 
-      before_action :authenticate_user!
+      before_action :require_not_suspended!
```
Comment (ThibG) [inline app/controllers/settings/two_factor_authentication/confirmations_controller.rb:10]: As I said in my previous comment, I think it's probably fine to allow changing 2FA settings for suspended users.

File: app/controllers/settings/two_factor_authentication/otp_authentication_controller.rb [hunk 7]
```diff
@@ -5,14 +5,12 @@ module TwoFactorAuthentication
     class OtpAuthenticationController < BaseController
       include ChallengableConcern
 
-      layout 'admin'
+      skip_before_action :require_functional!
 
-      before_action :authenticate_user!
+      before_action :require_not_suspended!
```
Comment (ThibG) [inline app/controllers/settings/two_factor_authentication/otp_authentication_controller.rb:10]: As I said in my previous comment, I think it's probably fine to allow changing 2FA settings for suspended users.

File: app/controllers/settings/two_factor_authentication/recovery_codes_controller.rb [hunk 8]
```diff
@@ -5,13 +5,11 @@ module TwoFactorAuthentication
     class RecoveryCodesController < BaseController
       include ChallengableConcern
 
-      layout 'admin'
+      skip_before_action :require_functional!
 
-      before_action :authenticate_user!
+      before_action :require_not_suspended!
```
Comment (ThibG) [inline app/controllers/settings/two_factor_authentication/recovery_codes_controller.rb:10]: As I said in my previous comment, I think it's probably fine to allow changing 2FA settings for suspended users.

File: app/controllers/settings/two_factor_authentication/webauthn_credentials_controller.rb [hunk 9]
```diff
@@ -3,9 +3,9 @@
 module Settings
   module TwoFactorAuthentication
     class WebauthnCredentialsController < BaseController
-      layout 'admin'
+      skip_before_action :require_functional!
 
-      before_action :authenticate_user!
+      before_action :require_not_suspended!
```
Comment (ThibG) [inline app/controllers/settings/two_factor_authentication/webauthn_credentials_controller.rb:8]: As I said in my previous comment, I think it's probably fine to allow changing 2FA settings for suspended users.

File: app/controllers/settings/two_factor_authentication_methods_controller.rb [hunk 10]
```diff
@@ -4,14 +4,12 @@ module Settings
   class TwoFactorAuthenticationMethodsController < BaseController
     include ChallengableConcern
 
-    layout 'admin'
+    skip_before_action :require_functional!
 
-    before_action :authenticate_user!
+    before_action :require_not_suspended!
```
Comment (ThibG) [inline app/controllers/settings/two_factor_authentication_methods_controller.rb:9]: As I said in my previous comment, I think it's probably fine to allow changing 2FA settings for suspended users.

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]

File: app/controllers/settings/featured_tags_controller.rb (hunk 2 above)
Comment (Gargron) [reply app/controllers/settings/featured_tags_controller.rb:4]: The controller inherits from Settings::BaseController (which sets layout and authenticate_user!), which inherits from ApplicationController (which does require_functional_account!). So, this controller is actually only accessible to funtional accounts. Suspended users excluded.

Comment (Gargron) [reply app/controllers/settings/featured_tags_controller.rb:4]: I'll be honest my brain gets fried too trying to think of what pending, unconfirmed, suspended, disabled, memorialized users are supposed to be able to access. I thought uniting all of that into "functional" would simplify it, and so it was, up until we needed to customize some behaviours between them.

File: app/controllers/settings/exports/blocked_accounts_controller.rb [hunk 11]
```diff
@@ -2,7 +2,7 @@
 
 module Settings
   module Exports
-    class BlockedAccountsController < ApplicationController
+    class BlockedAccountsController < BaseController
```
Comment (ThibG) [inline app/controllers/settings/exports/blocked_accounts_controller.rb:5]: ok, if this inherits from `Settings::BaseController`, doesn't that still `require_functional!`?

File: app/controllers/settings/exports/blocked_domains_controller.rb [hunk 12]
```diff
@@ -2,7 +2,7 @@
 
 module Settings
   module Exports
-    class BlockedDomainsController < ApplicationController
+    class BlockedDomainsController < BaseController
```
Comment (ThibG) [inline app/controllers/settings/exports/blocked_domains_controller.rb:5]: ok, if this inherits from Settings::BaseController, doesn't that still require_functional!?

File: app/controllers/settings/exports/following_accounts_controller.rb [hunk 13]
```diff
@@ -2,7 +2,7 @@
 
 module Settings
   module Exports
-    class FollowingAccountsController < ApplicationController
+    class FollowingAccountsController < BaseController
```
Comment (ThibG) [inline app/controllers/settings/exports/following_accounts_controller.rb:5]: ok, if this inherits from Settings::BaseController, doesn't that still require_functional!?

File: app/controllers/settings/exports/lists_controller.rb [hunk 14]
```diff
@@ -2,7 +2,7 @@
 
 module Settings
   module Exports
-    class ListsController < ApplicationController
+    class ListsController < BaseController
```
Comment (ThibG) [inline app/controllers/settings/exports/lists_controller.rb:5]: ok, if this inherits from Settings::BaseController, doesn't that still require_functional!?

File: app/controllers/settings/exports/muted_accounts_controller.rb [hunk 15]
```diff
@@ -2,7 +2,7 @@
 
 module Settings
   module Exports
-    class MutedAccountsController < ApplicationController
+    class MutedAccountsController < BaseController
```
Comment (ThibG) [inline app/controllers/settings/exports/muted_accounts_controller.rb:5]: ok, if this inherits from Settings::BaseController, doesn't that still require_functional!?

File: app/controllers/settings/featured_tags_controller.rb (hunk 2 above)
Comment (ThibG) [reply app/controllers/settings/featured_tags_controller.rb:4]: oh ok, it does make sense here

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]

File: app/controllers/settings/exports/blocked_accounts_controller.rb (hunk 11 above)
Comment (Gargron) [reply app/controllers/settings/exports/blocked_accounts_controller.rb:5]: `ExportsControllerConcern` adds `skip_before_action :require_functional!`

Comment (ThibG) [reply app/controllers/settings/exports/blocked_accounts_controller.rb:5]: Right! Sorry about that.

File: app/controllers/settings/sessions_controller.rb (hunk 5 above)
Comment (ThibG) [reply app/controllers/settings/sessions_controller.rb:6]: Well, that's the only action supported by `destroy`, so…

**Review by ClearlyClaire — APPROVED** [review]

**Pull request merged**